File: brainfuck/bf.lua
local INC = 0
local MOVE = 1
local PRINT = 3
local LOOP = 4

local function Tape()
  local pos = 1
  local data = {0}

  local get = function()
    return data[pos]
  end

  local inc = function(x)
    data[pos] = data[pos] + x
  end

  local move = function(x)
    local length = #data
    pos = pos + x
    for i = length + 1, pos do
      data[i] = 0
    end
  end

  return {
    get = get,
    inc = inc,
    move = move,
  }
end

local function Printer(quiet)
  local sum1 = 0
  local sum2 = 0

  local function print(n)
    if quiet then
      sum1 = (sum1 + n) % 255
      sum2 = (sum2 + sum1) % 255
    else
      io.write(string.char(n))
      io.flush()
    end
  end

  local function get_checksum()
    local bit32 = require "bit32"
    return bit32.bor(bit32.lshift(sum2, 8), sum1)
  end

  return {
    print = print,
    get_checksum = get_checksum,
    quiet = quiet
  }
end

local function Brainfuck(text, p)
  local function parse(source, i)
    local res = {}
    while i <= source:len() do
      local c = source:sub(i, i)
      if c == "+" then
        table.insert(res, {INC, 1})
      elseif c == "-" then
        table.insert(res, {INC, -1})
      elseif c == ">" then
        table.insert(res, {MOVE, 1})
      elseif c == "<" then
        table.insert(res, {MOVE, -1})
      elseif c == "." then
        table.insert(res, {PRINT})
      elseif c == "[" then
        local loop_code
        loop_code, i = parse(source, i+1)
        table.insert(res, {LOOP, loop_code})
      elseif c == "]" then
        break;
      end
      i = i + 1
    end
    return res, i
  end

  local function _run(program, tape)
    for i = 1, #program do
      local op = program[i]
      local operator = op[1]
      if operator == INC then
        tape.inc(op[2])
      elseif operator == MOVE then
        tape.move(op[2])
      elseif operator == LOOP then
        while tape.get() > 0 do
          _run(op[2], tape)
        end
      elseif operator == PRINT then
        p.print(tape.get())
      end
    end
  end

  local function run()
    _run(parse(text, 1), Tape())
  end

  return {
    run = run
  }
end

local function notify(msg)
  local socket = require "socket"
  socket.protect(function()
    local c = socket.try(socket.connect("localhost", 9001))
    local try = socket.newtry(function() c:close() end)
    try(c:send(msg))
    c:close()
  end)()
end

local function verify()
  text = [[
  ++++++++[>++++[>++>+++>+++>+<<<<-]>+>+>->>+[<]<-]>>.>
  ---.+++++++..+++.>>.<-.<.+++.------.--------.>>+.>++.
  ]]
  local p_left = Printer(true)
  Brainfuck(text, p_left).run()
  local left = p_left.get_checksum()

  local p_right = Printer(true)
  local str = "Hello World!\n"
  for i = 1, #str do
    p_right.print(string.byte(str, i))
  end
  local right = p_right.get_checksum()

  if left ~= right then
    print(string.format("%d != %d", left, right))
    os.exit(1)
  end
end

(function(arg)
    verify()
    local f = io.open(arg[1])
    local text = f:read("*a")
    f:close()
    local p = Printer(os.getenv("QUIET"))

    local compiler = type(jit) == 'table' and "Lua/luajit" or "Lua"
    local getpid = require 'posix.unistd'.getpid
    notify(string.format("%s\t%d", compiler, getpid()))
    Brainfuck(text, p).run()
    notify("stop")

    if p.quiet then
      print(string.format("Output checksum: %d", p.get_checksum()))
    end
end)(arg)

File: brainfuck/bf.ex
defmodule BF.Tape do
  def init() do
    {:atomics.new(30_000, []), 1, 1}
  end

  def current({data, pos, _max_pos}) do
    :atomics.get(data, pos)
  end

  def inc({data, pos, max_pos}, x) when x == -1 or x == 1 do
    :atomics.add(data, pos, x)

    {data, pos, max_pos}
  end

  def move({data, pos, max_pos}, x) when x == -1 or x == 1 do
    new_pos = pos + x

    if new_pos <= max_pos do
      {data, new_pos, max_pos}
    else
      :atomics.put(data, new_pos, 0)
      {data, new_pos, new_pos}
    end
  end
end

defmodule Printer do
  @enforce_keys [:quiet]
  defstruct [:quiet, sum1: 0, sum2: 0]

  def print(n, p) do
    if p.quiet do
      new_sum1 = rem(p.sum1 + n, 255)
      new_sum2 = rem(p.sum2 + new_sum1, 255)
      %Printer{p | sum1: new_sum1, sum2: new_sum2}
    else
      :ok = IO.binwrite([n])
      :file.sync(:stdout)
      p
    end
  end

  def get_checksum(p) do
    import Bitwise
    (p.sum2 <<< 8) ||| p.sum1
  end
end

defmodule BF do
  alias BF.Tape, as: Tape

  def parse(source) do
    {result, []} = parse(to_charlist(source), [])
    result
  end

  defp parse([], acc) do
    {Enum.reverse(acc), []}
  end

  defp parse([x | xs], acc) do
    case x do
      ?+ -> parse(xs, [{:inc, 1} | acc])
      ?- -> parse(xs, [{:inc, -1} | acc])
      ?> -> parse(xs, [{:move, 1} | acc])
      ?< -> parse(xs, [{:move, -1} | acc])
      ?. -> parse(xs, [{:print, nil} | acc])
      ?[ ->
        {loop_code, new_xs} = parse(xs, [])
        parse(new_xs, [{:loop, loop_code} | acc])
      ?] -> {Enum.reverse(acc), xs}
      _   -> parse(xs, acc)
    end
  end

  def run([], tape, p) do
    {tape, p}
  end

  def run(program = [op | ops], tape, p) do
    case op do
      {:inc, x} -> run(ops, Tape.inc(tape, x), p)
      {:move, x} -> run(ops, Tape.move(tape, x), p)
      {:print, nil} ->
        p = Printer.print(Tape.current(tape), p)
        run(ops, tape, p)
      {:loop, loop_code} ->
        if Tape.current(tape) == 0 do
          run(ops, tape, p)
        else
          {tape, p} = run(loop_code, tape, p)
          run(program, tape, p)
        end
    end
  end
end

defmodule Benchmark do
  def notify(msg) do
    with {:ok, socket} <- :gen_tcp.connect('localhost', 9001, []) do
      :gen_tcp.send(socket, msg)
      :gen_tcp.close(socket)
    end
  end

  def verify do
    text = """
    ++++++++[>++++[>++>+++>+++>+<<<<-]>+>+>->>+[<]<-]>>.>
    ---.+++++++..+++.>>.<-.<.+++.------.--------.>>+.>++.
    """
    left = Printer.get_checksum(
      elem(BF.parse(text)
      |> BF.run(BF.Tape.init(), %Printer{quiet: true}), 1))

    right = Printer.get_checksum(
      Enum.reduce(
        String.to_charlist("Hello World!\n"),
        %Printer{quiet: true},
        &Printer.print/2))
    if left != right do
      IO.puts(:stderr, "#{left} != #{right}")
      exit(:shutdown)
    end
  end

  def run do
    verify()
    [filename] = System.argv()
    text = File.read!(filename)

    notify("Elixir\t#{System.pid()}")

    p = text
      |> BF.parse()
      |> BF.run(BF.Tape.init(), %Printer{quiet: System.get_env("QUIET")})
      |> elem(1)

    notify("stop")

    if p.quiet do
      IO.puts("Output checksum: #{Printer.get_checksum(p)}")
    end
  end
end

Benchmark.run()
